---
title: "Multi-Omic project"
date: September 4, 2023
output: 
  BiocStyle::html_document:
      toc: true
package: methylkey
params:
  idat: path_to_idat
  sampleSheet : path_to_samplesheet_file
  nalimit : 0.2
  pval : 0.02
  normalize : "Funnorm"
  output : output_directory
---

```{R setup, echo = FALSE, message=FALSE, warning = FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning = FALSE)
library(methylkey)#v0.1.1.900
library(tidyverse)#v2.0.0
library(DT)#v0.28
library(minfi)#v1.40.0
```

# loading data

```{r}
samplesheet_file<-data.frame( Basename=gsub("_Red.idat","",list.files(params$idat,pattern = "Red") )) %>%
  mutate(samples=Basename) %>% separate(Basename,into=c("Sentrix_ID","Sentrix_position"),remove = F )

```  

```{r, warning=FALSE}
RGset<-minfi::read.metharray.exp(base = params$idat, targets=samplesheet_file, force=TRUE)
betas <- minfi::getBeta(RGset)
plateform=methylkey::getPlateform(betas)
cat( paste("plateform =", plateform) )
```

# QC {.tabset}

## getSex checks

```{r}
GMsetEx <- mapToGenome(RGset) 
estSex  <- getSex(GMsetEx)
pdata<-sampleSheet
pdata$sex<-estSex$predictedSex
rownames(pdata)<-colnames(betas)

samplesheet_file <- samplesheet_file %>% 
               mutate( predictedSex = estSex$predictedSex)
               
               
samplesheet_file %>% dplyr::select(any_of(c("samples","Basename","Gender","predictedSex","Age"))) %>% datatable()
```

## Color Channels Intensities

```{r}
nsamp=nrow(samplesheet_file_file)
ylab<-"log2 intensity of both green and red channel"
par(xaxt='n')
boxplot(log2(getRed(RGset)+1), col = "firebrick", boxwex = 0.25, at= 1:nsamp - 0.2, ylab=ylab, labels=samplesheet_file$samples, cex=0.4)
boxplot(log2(getGreen(RGset)+1), col = "steelblue4", boxwex = 0.25, at= 1:nsamp + 0.2, axis=F , add=T, cex=0.4)
par(xaxt='s')
axis(1, at=1:nsamp, labels=samplesheet_file_file$samples, tick=TRUE, las=2, cex.axis=0.6)

boxplot(log2(getRed(RGset)+1), main="Red channel", col = "firebrick", las=2, cex.axis=0.6, labels=samplesheet_file$samples)
boxplot(log(getGreen(RGset)+1), main="Green channel", col = "steelblue4", las=2, cex.axis=0.6, labels=samplesheet_file$samples)
```

## densityPlot

```{r}
color=rep(c('firebrick','steelblue4'),length(unique(samplesheet_file$Sentrix_ID)))
  
densityPlot(betas, 
	            main="Density plot", 
	            xlab="Beta", 
	            pal=color
	         )

densityBeanPlot(betas, 
	           sampGroups=samplesheet_file$Sentrix_ID, 
	           sampNames=samplesheet_file$samples, 
	           main="Density Bean plot", 
	           pal=color,
	           numPositions=10000
	       )
```

## MDS Plot

```{r}
mdsPlot( betas,
          sampGroups=samplesheet_file_file$Sentrix_ID, 
          pal=color,
          numPositions=10000,
          main="MDS plot"
      )
```


## Estimate sample-specific quality control (QC) for methylation data

```{r}
MSet <- preprocessRaw(RGset)
MSet <- fixMethOutliers(MSet)
qc <- getQC(MSet)
plotQC(qc)
```

## NA values

```{r}
pvalues  <- minfi::detectionP(RGset)
colnames(pvalues)<-samplesheet_file$samples
data.frame(pvalues) %>% rownames_to_column("probes") %>% gather("samples","pvalue", -probes) %>% group_by(samples) %>% summarise(count=sum(pvalue>0.02)) %>% ggplot(aes(x=samples, y=count)) + geom_col() + 
  xlab('')+ ylab('Counts of missing probes out of ~850k')+theme_classic(base_size = 8) +theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1))
```

## Normalisation {.tabset}

```{r}
isna<-is.na(betas)

if(params$normalize=="Funnorm") {	MSet<-preprocessFunnorm(RGset, sex=estSex$predictedSex)   }
if(params$normalize=="Illumina"){	MSet<-preprocessIllumina(RGset)                           }
if(params$normalize=="Noob")    {	MSet<-preprocessNoob(RGset)	                              }
if(params$normalize=="Quantile"){	MSet<-preprocessQuantile(RGset, sex=estSex$predictedSex)  }
if(params$normalize=="SWANoob") { MSet<-preprocessSWAN(RGset, preprocessNoob(RGset))        }
if(params$normalize=="SWAN")    { MSet<-preprocessSWAN(RGset, preprocessRaw(RGset))         }
betas=getBeta(MSet)

#After normalization NA values are replace by values close to 0. This restore the NA status.
isna<-isna[ match(rownames(betas), rownames(isna)), ]
betas[ which(isna) ]<-NA 
any(is.na(betas))
```

## densityPlot

```{r}
n<-max(3,length( unique( samplesheet_file$Sentrix_ID) ))
densityPlot(betas, 
	            #sampGroups = samplesheet_file$Sentrix_ID, 
	            main="Density plot", 
	            xlab="Betas", 
	            pal=color 
	         )

densityBeanPlot(betas, 
	           sampGroups=samplesheet_file$Sentrix_ID, 
	           sampNames=samplesheet_file$samples, 
	           main="Density Bean plot", 
	           pal=color,
	           numPositions=10000
	       )
```

# Filtering {.tabset}

## filters

```{r}
# remove betas with low quality ( high pvalue )
betas[ pvalues[rownames(betas),] > params$pval ] <- NA
print( paste0( "Low quality probes :", sum(pvalues > params$pval), " low quality probes replaced by NA"  ) )
```

```{r}
# Filter crossReactive, snp and xy probes
probes  <- CpGexcl(plateform=plateform)
betas   <- betas[ ! rownames(betas) %in% probes,] 
print( paste0( "Filtered probes : ", length(probes), " filtered from blacklists"  ) )
```

```{r}
# Filter probes with too many na values
naprobes<-CpGNAexcl( betas,params$nalimit )
betas<-betas[ ! rownames(betas) %in% naprobes,]
print( paste0( "Filtered probes : ", length(naprobes), " with proportion of NA values > ", params$nalimit    ) )
```


## filtered probes

```{r}
print(probes)
print(naprobes)
#Reset the NAs to 0 for dowstream analyses
betas[is.na(betas)]<-0
any (is.na(betas))
```

## MDS Plot

```{r}
mdsPlot( betas,
          sampGroups=samplesheet_file$Sentrix_ID, 
          #sampNames=samplesheet_file$samples, 
          pal=color,
          numPositions=10000,
          main="MDS plot"
      )
```
##Checking for Batch effects
```{r}
#
mval=beta2m(betas)
any(is.na(betas))
mval[is.infinite(mval)]<-0
probes <- names(which( rowSums(apply(mval, 2, is.infinite )) >0))
mval1=bc_sva(mval = mval, pdata=samplesheet_file, model='~Batch')

```


```{r}
save(samplesheet_file,mval,mval1,betas,RGset,MSet,file=paste0( params$output, "_", params$normalize, "_", Sys.Date(),".rda" ) )

```

# sessionInfo

```{r}
sessionInfo()
```
